lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1613,30 +1624,72 @@ class HSAQueue final : public KalmarQueue
-                    }
-                    // wait on valid futures only
-                    std::shared_future<void>* future = asyncOp->getFuture();
-                    if (future && future->valid()) {
-                        future->wait();
-                    }
+        for (int i = lastWaitOp-1; i >= 0;  i--) {
+            if (asyncOps[i] != nullptr) {
+                auto asyncOp = asyncOps[i];
+                // we only drain starting at first found marker
+                if (drainingQueue_ && asyncOp->getCommandKind() != hcCommandMarker) continue;
+#if !defined(NDEBUG)
+                if (!foundFirstValidOp) {
+                    hsa_signal_t sig =  *(static_cast <hsa_signal_t*> (asyncOp->getNativeHandle()));
+                    assert(sig.handle != 0);
+                    foundFirstValidOp = true;
+                }
+#endif
+                // wait on valid futures only
+                std::shared_future<void>* future = asyncOp->getFuture();
+                if (future && future->valid()) {
+                    future->wait();
                 }
             }
-
-            // clear async operations table
+            else {
+                first_nullptr = i;
+                break; /* as soon as an op is found to be nullptr, we know the older ones are also */
+            }
+        }
+        // clear async operations table
+        if (drainingQueue_ && first_nullptr >= 0) {
+            /* Did the waited op become null? This can happen due to future->wait() above. */
+            if (asyncOps[lastWaitOp-1] == nullptr) {
+                first_nullptr = lastWaitOp-1;
+            }
+            /* we can only safely clear from the first nullptr and older */
+            if (first_nullptr >= 0) {
+                // We can do something faster than
+                // asyncOps.erase(asyncOps.begin(), asyncOps.begin() + first_nullptr + 1);
+                // because erasing from front of vectors will repeatedly move remaining elements.
+                // We can also do better than the improved
+                // asyncOps.erase(std::remove(asyncOps.begin(), asyncOps.begin()+first_nullptr+1, nullptr), asyncOps.end());
+                // because we know that all elements from 0..first_nullptr are nullptrs,
+                // so we can avoid the redundant checks of each element against nullptr.
+                for (size_t i=first_nullptr+1,new_i=0,end=asyncOps.size(); i<end; ++i) {
+                    asyncOps[new_i++] = std::move(asyncOps[i]);
+                }
+                asyncOps.resize(asyncOps.size() - first_nullptr - 1);
+                asyncOps_offset += first_nullptr + 1;
+                DBOUTL(DB_RESOURCE, "asyncOps=" << &asyncOps << " * erasing " << first_nullptr+1 << " ops. New size " << asyncOps.size());
+            }
+        }
+        else if (drainingQueue_) {
+            /* we were not able to find a suitable op to wait on */
+            drainingQueueFailed_ = true;
+        }
+        else {
```

Comment:
there are 3 checks in this block:
```
if (drainingQueue_ && first_nullptr >= 0) {
} else if (drainingQueue_) {
} else {
}
```

would it be better to re-organize it to:

```
if (drainingQueue_) {
  if (first_nullptr >= 0) {
  } else {
  }
} else {
}
```

so perhaps it's better comprehensible?